state-sync/storage-service/server/src/tests.rs

```diff
@@ -13,57 +13,192 @@ use aptos_types::{
     StorageServiceError, StorageServiceMessage, StorageServiceRequest, StorageServiceResponse,
     TransactionOutputsWithProofRequest, TransactionsWithProofRequest,
 };
 
-// TODO(joshlind): Expand these test cases to better test storage interaction
-// and functionality. This will likely require a better mock db abstraction.
-
 /// Various test constants for storage
-const FIRST_TXN_OUTPUT_VERSION: u64 = 20;
-const FIRST_TXN_VERSION: u64 = 10;
-const LAST_EPOCH: u64 = 10;
-const LAST_TXN_VERSION: u64 = 100;
-const NUM_ACCOUNTS_AT_VERSION: u64 = 1000;
 const PROTOCOL_VERSION: u64 = 1;
-const STATE_PRUNE_WINDOW: u64 = 50;
+
+#[tokio::test]
+async fn test_cachable_requests_eviction() {
+    // Create test data
+    let max_lru_cache_size = StorageServiceConfig::default().max_lru_cache_size;
+    let version = 101;
+    let start_account_index = 100;
+    let end_account_index = 199;
+    let state_value_chunk_with_proof = StateValueChunkWithProof {
+        first_index: start_account_index,
+        last_index: end_account_index,
+        first_key: HashValue::random(),
+        last_key: HashValue::random(),
+        raw_values: vec![],
+        proof: SparseMerkleRangeProof::new(vec![]),
+        root_hash: HashValue::random(),
+    };
+
+    // Create the mock db reader
+    let mut db_reader = create_mock_db_reader();
+    let mut expectation_sequence = Sequence::new();
+    db_reader
+        .expect_get_state_leaf_count()
+        .times(max_lru_cache_size as usize)
+        .with(always())
+        .returning(move |_| Ok(165));
+    for _ in 0..2 {
+        let state_value_chunk_with_proof_clone = state_value_chunk_with_proof.clone();
+        db_reader
+            .expect_get_state_value_chunk_with_proof()
+            .times(1)
+            .with(
+                eq(version),
+                eq(start_account_index as usize),
+                eq((end_account_index - start_account_index + 1) as usize),
+            )
+            .return_once(move |_, _, _| Ok(state_value_chunk_with_proof_clone))
+            .in_sequence(&mut expectation_sequence);
+    }
+
+    // Create the storage client and server
+    let (mut mock_client, service, _) = MockClient::new(Some(db_reader));
+    tokio::spawn(service.start());
+
+    // Process a request to fetch an account states chunk. This should cache and serve the response.
+    for _ in 0..2 {
+        let request = StorageServiceRequest::GetAccountStatesChunkWithProof(
+            AccountStatesChunkWithProofRequest {
+                version,
+                start_account_index,
+                end_account_index,
+            },
+        );
+        let _ = mock_client.send_request(request).await.unwrap();
+    }
+
+    // Process enough requests to evict the previously cached response
+    for version in 0..max_lru_cache_size {
+        let request = StorageServiceRequest::GetNumberOfAccountsAtVersion(version);
+        let _ = mock_client.send_request(request).await.unwrap();
```

Comment:
where are you checking to make sure the cache is actually evicted?

Reply:
This is done by the `.expect_get_state_value_chunk_with_proof()` call on the `db_reader` above. We call this method twice (in the loop), telling the mock that we expect two calls to the mock storage. The first call should be for the data (that is put in the cache). The second happens when the data is evicted and we make a call for it again. Does that make sense?